python/ray/train/bagua.py

```diff
@@ -0,0 +1,184 @@
+from dataclasses import dataclass
+import logging
+import os
+
+from typing import Optional
+
+import ray
+from ray.train.torch import (
+    TorchAccelerator,
+    TorchBackend,
+)
+
+from ray.train.backend import BackendConfig
+from ray.train.session import get_accelerator, set_accelerator
+from ray.train.worker_group import WorkerGroup
+from ray.train.utils import get_address_and_port
+from ray.util import PublicAPI
+
+import torch
+import torch.distributed as dist
+from torch.utils.data import DistributedSampler
+
+import bagua.torch_api
+from bagua.torch_api.algorithms import gradient_allreduce
+
+try:
+    from torch.profiler import profile
+except ImportError:
+    profile = None
+
+logger = logging.getLogger(__name__)
+
+
+def _get_torch_distributed_sampler(dataset, shuffle):
+    return DistributedSampler(
+        dataset,
+        num_replicas=bagua.torch_api.get_world_size(),  # equivalent to ray.train.size()
+        rank=bagua.torch_api.get_rank(),  # equivalent to ray.train.world_rank()
+        shuffle=shuffle,
+    )
+
+
+class BaguaAccelerator(TorchAccelerator):
+    def __init__(self, amp: bool = False):
+        super().__init__(amp=amp)
+
+    def prepare_model(
+        self,
+        model: torch.nn.Module,
+        optimizer: torch.optim.Optimizer,
+        algorithm: bagua.torch_api.algorithms.Algorithm,
+    ) -> torch.nn.Module:
+
+        device = self.get_device()
+        model = model.to(device)
+
+        if torch.cuda.is_available():
+            torch.cuda.set_device(device)
+
+        if self.amp_is_enabled:
+            # move wrap_forward() and model_get_state() to a new function
+            model = self._patch_model_forward_and_state(model)
+
+        logger.info("Wrapping provided model in BaguaDDP.")
+        # we need the optimizer when preparing the model
+        # with_bagua() returns a BaguaModule not torch.nn.Module
+        model = model.with_bagua([optimizer], algorithm)
```

Comment:
Where is `with_bagua` defined? It seems like `model` is just a standard `torch.nn.Module`?

Reply:
Also do we need to wrap in bagua.torch_api.DistributedDataParallel anywhere? Or is this what `with_bagua` is doing?

Reply:
The method [`with_bagua()`](https://github.com/BaguaSys/bagua/blob/master/bagua/torch_api/distributed.py#L53-L59) is patched to `torch.nn.Module`. When calling `type(model)`, both return `<class 'torch.nn.modules.SomeClass>`. However, the actual behaviors are not the same. For example, on `torch.nn.Module` there is no problem with `copy.deepcopy(model)`. On the other hand,  this causes an error `TypeError: cannot pickle 'Stream' object` if the model is returned by `with_bagua()`. I will look more into this issue.

Regarding the DDP part, it seems Bagua creates its own DDP instance using [`BaguaDistributedDataParallel`](https://github.com/BaguaSys/bagua/blob/32be6eba7e6e53ed567f602297c208d8fff1f3d7/bagua/torch_api/data_parallel/bagua_distributed.py#L27), and it's done in `with_bagua()` ([code](https://github.com/BaguaSys/bagua/blob/master/bagua/torch_api/distributed.py#L105-L112)).